### PR TITLE
Replace XML descriptors for SRE with DynamicDependency attributes

### LINKED ISSUE #38692: Mono's ILLink Descriptor is preserving all the fields on Reflection.Emit types
See the `preserve="fields"` on these types:

https://github.com/dotnet/runtime/blob/d89772ae2c8ce4f7e0bb7af4891619ff759ee1c0/src/mono/netcore/System.Private.CoreLib/src/ILLink/System.Private.CoreLib.xml#L617-L648

When `Reflection.Emit` is no longer used in an app, `AssemblyBuilder` still gets pulled in because `AssemblyLoadContext::GetRuntimeAssembly` is checking for `AssemblyBuilder`:

```
	Field:System.Reflection.Emit.ILExceptionInfo[] System.Reflection.Emit.ILGenerator::ex_handlers size: 0
	| TypeDef:System.Reflection.Emit.ILGenerator [31 deps]
	| Field:System.Reflection.Emit.ILGenerator System.Reflection.Emit.MethodBuilder::ilgen [1 deps]
	| TypeDef:System.Reflection.Emit.MethodBuilder [48 deps]
	| TypeSpec:System.Reflection.Emit.MethodBuilder[] [3 deps]
	| Field:System.Reflection.Emit.MethodBuilder[] System.Reflection.Emit.TypeBuilder::methods [1 deps]
	| TypeDef:System.Reflection.Emit.TypeBuilder [74 deps]
	| TypeSpec:System.Reflection.Emit.TypeBuilder[] [3 deps]
	| Field:System.Reflection.Emit.TypeBuilder[] System.Reflection.Emit.ModuleBuilder::types [1 deps]
	| TypeDef:System.Reflection.Emit.ModuleBuilder [37 deps]
	| TypeSpec:System.Reflection.Emit.ModuleBuilder[] [1 deps]
	| Field:System.Reflection.Emit.ModuleBuilder[] System.Reflection.Emit.AssemblyBuilder::modules [1 deps]
	| TypeDef:System.Reflection.Emit.AssemblyBuilder [35 deps]
	| Method:System.Reflection.RuntimeAssembly System.Runtime.Loader.AssemblyLoadContext::GetRuntimeAssembly(System.Reflection.Assembly) [1 deps]
	| Method:System.Reflection.RuntimeAssembly System.Runtime.Loader.AssemblyLoadContext::InvokeResolveEvent(System.ResolveEventHandler,System.Reflection.RuntimeAssembly,System.String) [3 deps]
	| Method:System.Reflection.RuntimeAssembly System.Runtime.Loader.AssemblyLoadContext::OnResourceResolve(System.Reflection.RuntimeAssembly,System.String) [1 deps]
	| Other:resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
```

https://github.com/dotnet/runtime/blob/acd8e5d4c32b42e0d0ec457e0347f1d9d1eefada/src/mono/netcore/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.Mono.cs#L170-L177

Since, `AssemblyBuilder`s fields are preserved, it brings in `ModuleBuilder`, and so on.

We should see if we can remove the `preserve="fields"` on these types. I traced it back to the original commit of the linker into the mono repo: https://github.com/mono/mono/commit/00ebb8d9189f76bbae8e1556ac999553d856f72d, but that didn't tell me why these fields need to be preserved.

@marek-safar @vargaz @vitek-karas @MichalStrehovsky

### PR DESCRIPTION
**Pull request by marek-safar**
Fixes #38692

### CODE REVIEW

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/DerivedTypes.Mono.cs [hunk 1]
```diff
@@ -38,8 +38,11 @@ namespace System.Reflection.Emit
     [StructLayout(LayoutKind.Sequential)]
     internal abstract partial class SymbolType : TypeInfo
     {
-        internal Type m_baseType;
+#region Sync with MonoReflectionDerivedType in object-internals.h
+        private protected Type m_baseType;
+#endregion
 
+        [DynamicDependency(nameof(m_baseType))]  // Automatically keeps all previous fields too due to StructLayout
```
Comment (stephentoub) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/DerivedTypes.Mono.cs:45]: Any concerns with the robustness of this, if/when fields are reorganized?

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.Mono.cs [hunk 2]
```diff
@@ -56,9 +56,11 @@ public sealed partial class ConstructorBuilder : ConstructorInfo
         private Type[][]? paramModReq;
         private Type[][]? paramModOpt;
         private object? permissions;
-#pragma warning restore 169, 414
+#endregion
+
         internal bool finished;
 
+        [DynamicDependency(nameof(permissions))] // Automatically keeps all previous fields too due to StructLayout
```
Comment (stephentoub) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.Mono.cs:63]: Any concerns with the robustness of this, if/when fields are reorganized?

**Review by stephentoub — APPROVED** [review]
Nice.

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.Mono.cs [hunk 3]
```diff
@@ -34,14 +34,14 @@
 using System.Globalization;
 using System.Diagnostics.CodeAnalysis;
 using System.Runtime.InteropServices;
+using System.Diagnostics.CodeAnalysis;
```
Comment (stephentoub) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.Mono.cs:37]: Duplicate?

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.Mono.cs (hunk 2 above)
Comment (marek-safar) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.Mono.cs:62]: Not really. I think there are enough comments to catch that and right now we keep all fields

**Review by vargaz — APPROVED** [review]

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**